WalletWasabi/Stores/IndexStore.cs

```diff
@@ -67,10 +80,74 @@ public async Task InitializeAsync(CancellationToken cancellationToken)
 		{
 			cancellationToken.ThrowIfCancellationRequested();
 
+			// Migration code.
+			if (RunMigration)
+			{
+				MigrateToSqliteNoLock();
+			}
+
 			await InitializeFiltersNoLockAsync(cancellationToken).ConfigureAwait(false);
 		}
 	}
 
+	private void MigrateToSqliteNoLock()
+	{
+		int i = 0;
+
+		try
+		{
+			Logger.LogWarning("Migration of block filters to SQLite format is about to begin. Please wait a moment.");
+
+			Stopwatch stopwatch = Stopwatch.StartNew();
+
+			List<string> filters = new(capacity: 100_000);
+
+			using (FileStream fs = File.Open(OldIndexFilePath, FileMode.Open, FileAccess.Read, FileShare.Read))
+			using (BufferedStream bs = new(fs))
+			using (StreamReader sr = new(bs))
+			{
+				while (true)
+				{
+					i++;
+					string? line = sr.ReadLine();
```

Comment:
There are short living strings, so it's actually fine. GC likes objects which are not promoted to a next generation. If there's a large string then it goes to a LOH directly and stays there for a very long time. Just be sure that the string is small enough to go into the small object heap (what is the string length here?).

If you need to turn it to eleven then I would suggest to use `ReadBlock` to use a preallocated buffer, avoid `FromHexString` which allocates arrays by using a preallocated array too (all strings have the same length, right?). But that's an one time operation, so not sure that it's worth.

Another option here is parallelization through `Channel<T>` where one thread reads from the file and another one populates the database. It would minimize pauses between reads from the file and minimize pauses between writes to the database.

Reply:
.NET has a very cheap allocations due to its implementation which even is inlined (hot path only). There's are two pointers one of which is bumped till reaches the second, then a cold path is used which may trigger the GC and request a new page from the OS.

What isn't cheap is deallocations because the GC has to analyze references and determine what to kill and what to promote (: